### PR TITLE
POST /events/:id/pending-applicants/:userId/approve

### LINKED ISSUE #23: Implement `POST /events/:id/pending-applicants/:userId/approve`
## Tasks:

- Implement `POST /events/:id/pending-applicants/:userId/approve`
- Add documentation and unit testing.

### PR DESCRIPTION
**Pull request by yamanrajab90**
Fixes #23

### CODE REVIEW

File: src/controllers/__tests__/events.test.js [hunk 1]
```diff
@@ -150,4 +150,26 @@ describe("Testing events for routes doesn't require auth controls", () => {
...
+          expect(res.statusCode).toBe(404);
+          done();
+          return res.json('No event found');
+        }
+        expect(res.body.data).toEqual(events[0]);
+        expect(typeof res.body).toEqual('object');
+        done();
+        return events;
+      });
+  });
```
Comment (dilarafirtina) [inline src/controllers/__tests__/events.test.js:174]: Hafız detected that this test always passes so we need to change it a bit.

File: src/controllers/events.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
+const mongoose = require('mongoose');
 const Event = require('../models/event');
-const { Ngo } = require('../models/user');
+const { Ngo, User, Volunteer } = require('../models/user');
```
Comment (dilarafirtina) [inline src/controllers/events.js:3]: Great work at requiring all of them at once.

File: src/controllers/events.js [hunk 3]
```diff
@@ -29,3 +30,37 @@ exports.getRelatedEvents = async (req, res, next) => {
...
+  //     )
+  //   );
+  // }
+  const applicant = await Volunteer.findById(req.params.userId);
+  const applicantIndex = event.pendingApplicants.indexOf(applicant._id);
+  if (applicantIndex === -1) {
+    return next(
+      new ErrorResponse('This Volunteer did not apply for this event', 400)
+    );
+  }
```
Comment (dilarafirtina) [inline src/controllers/events.js:54]: Good way of checking array!

File: src/controllers/events.js [hunk 4]
```diff
@@ -29,3 +30,37 @@ exports.getRelatedEvents = async (req, res, next) => {
...
+    { _id: mongoose.Types.ObjectId(req.params.id) },
+    { $push: { approvedApplicants: applicant._id } }
+  );
+  await Event.findOneAndUpdate(
+    { _id: mongoose.Types.ObjectId(req.params.id) },
+    { $pull: { pendingApplicants: applicant._id } }
+  );
+  const updatedEvent = await Event.findById(req.params.id);
+  return res.status(200).json({ success: true, data: updatedEvent });
+};
```
Comment (dilarafirtina) [inline src/controllers/events.js:66]: Great way of implementing push and pull. I wasn't aware of this and was implementing like below. 
```
const event = await Event.findById(req.params.id)
event.pendingApplicants.push(req.params.userId
```)...etc

File: src/routes/events.js [hunk 5]
```diff
@@ -8,4 +8,11 @@ router.get('/:id', eventsControllers.getEvent);
 
 router.get('/:id/related-events', eventsControllers.getRelatedEvents);
 
+router.post(
+  '/:id/pending-applicants/:userId/approve',
+  // isAuth,
+  // isNgo,
```
Comment (dilarafirtina) [inline src/routes/events.js:14]: Yes, we need to add these middlewares, good catch.

File: swagger.json [hunk 6]
```diff
@@ -715,6 +715,57 @@
...
+                "400": {
+                  "description": "bad request",
+                  "example": "bad request"
+                }
+              }
+            }
+          }
+        }
+      }
+    },
```
Comment (dilarafirtina) [inline swagger.json:768]: Swagger will change and we can discuss it later.

**Review by dilarafirtina — APPROVED** [review]

File: src/controllers/__tests__/events.test.js (hunk 1 above)
Comment (yamanrajab90) [reply src/controllers/__tests__/events.test.js:174]: Thank you @dilarafirtina for all your comments, and this one especially cuz it is crucial.

**Pull request closed without merging**